packages/ui/src/common/components/forms/DatePicker/DatePicker.tsx

```diff
@@ -0,0 +1,182 @@
+import { addMonths, addWeeks, addYears, isAfter, isBefore, isEqual, startOfMonth, startOfToday } from 'date-fns'
+import React, { useEffect, useMemo, useRef, useState } from 'react'
+import styled from 'styled-components'
+
+import { ButtonSecondary, FilterButtons } from '@/common/components/buttons'
+import { Shadows } from '@/common/constants'
+import { DateRange, PartialDateRange } from '@/common/types/Dates'
+import { earliest, fromRange, latest, toDDMMYY } from '@/common/utils/dates'
+
+import { Calendar } from '../Calendar'
+import { InputComponent, InputText } from '../InputComponent'
+
+interface DatePickerProps {
+  title?: string
+  value: PartialDateRange
+  withinDates?: PartialDateRange
+  onApply?: () => void
+  onClear?: () => void
+  onChange: (value: PartialDateRange) => void
+}
+
+export const DatePicker = ({ title, value, withinDates, onApply, onClear, onChange }: DatePickerProps) => {
+  const placeholder = '__/__/__'
+  const { start, end } = fromRange(value)
+  const dateString = `${toDDMMYY(start) ?? placeholder} - ${toDDMMYY(end) ?? placeholder}`
+
+  const container = useRef<HTMLDivElement>(null)
+  const [isOpen, toggleOpen] = useState(false)
+
+  useEffect(() => {
+    if (!isOpen) return
+
+    const closePopup = (evt: MouseEvent) => {
+      evt.stopPropagation()
+      if (!evt.target) return
+
+      const target = evt.target as Node
+      const clickedOutside = !container.current?.contains(target) ?? false
+      clickedOutside && toggleOpen(false)
+    }
+
+    window.addEventListener('mousedown', closePopup)
+    return () => window.removeEventListener('mousedown', closePopup)
+  }, [isOpen])
+
+  const apply = () => {
+    onApply?.()
+    toggleOpen(false)
+  }
+
+  return (
+    <DatePickerContainer ref={container} onMouseDown={() => !isOpen && toggleOpen(true)}>
+      <InputComponent tight inputWidth="xs">
+        <InputText placeholder="-" value={dateString} readOnly />
+        {isOpen && (
+          <DatePickerPopup>
+            <DatePickerCalendars value={value} within={withinDates} onChange={onChange} />
+            <DatePickerFooter>
+              <FilterButtons onApply={apply} onClear={onClear} />
+            </DatePickerFooter>
+          </DatePickerPopup>
+        )}
+      </InputComponent>
+    </DatePickerContainer>
+  )
+}
+
+interface DatePickerCalendarsProps {
+  value: PartialDateRange
+  within?: PartialDateRange
+  onChange: (value: PartialDateRange) => void
+}
+const DatePickerCalendars = ({ value, within, onChange }: DatePickerCalendarsProps) => {
+  const { start, end } = fromRange(value)
+
+  const today = useMemo(() => startOfToday(), [])
+  const thisMonth = useMemo(() => startOfMonth(today), [])
+  const [monthL, setMonthL] = useState(start ? startOfMonth(start) : addMonths(thisMonth, -1))
+  const [monthR, setMonthR] = useState(end ? startOfMonth(end) : thisMonth)
+
+  const selectionStrategy = (k1: keyof DateRange) => (day: Date) => {
```

Comment:
The `k1` is again too cryptic. While reading I have no idea what this might mean ;)

However, I see that there's `as`  cast used a lot so maybe the `Date` type might be different :thinking: 

For now, let's keep it this way, I'll have another look when you split the PR to a smaller one.

Reply:
Sorry this one was a bit tricky to name, but with `k1` I didn't try very hard :facepalm: 

The multiple casts come from typescript not supporting union types (in this case `'start' | 'end'`) as object keys. Instead it will fallback to inferring { [k: string]: Date } but when casting the key to `'start' | 'end'` it outputs: `An index signature parameter type cannot be a union type`